docs/memory_layout.md

```diff
@@ -0,0 +1,230 @@
+| mat2x2\<int8>[]  | <kbd>[bb--][----][----][----]<br/>[bb--][----][----][----]                                                          </kbd> |
+| mat2x2\<fp16>[]  | <kbd>[bbbb][----][----][----]<br/>[bbbb][----][----][----]                                                          </kbd> |
+| mat2x2\<float>[] | <kbd>[bbbb][bbbb][----][----]<br/>[bbbb][bbbb][----][----]                                                          </kbd> |
+| mat2x3\<float>[] | <kbd>[bbbb][bbbb][bbbb][----]<br/>[bbbb][bbbb][bbbb][----]                                                          </kbd> |
+| mat2x4\<float>[] | <kbd>[bbbb][bbbb][bbbb][bbbb]<br/>[bbbb][bbbb][bbbb][bbbb]                                                          </kbd> |
+| mat3x2\<float>[] | <kbd>[bbbb][bbbb][----][----]<br/>[bbbb][bbbb][----][----]<br/>[bbbb][bbbb][----][----]                             </kbd> |
+| mat4x2\<float>[] | <kbd>[bbbb][bbbb][----][----]<br/>[bbbb][bbbb][----][----]<br/>[bbbb][bbbb][----][----]<br/>[bbbb][bbbb][----][----]</kbd> |
+| mat4x3\<float>[] | <kbd>[bbbb][bbbb][bbbb][----]<br/>[bbbb][bbbb][bbbb][----]<br/>[bbbb][bbbb][bbbb][----]<br/>[bbbb][bbbb][bbbb][----]</kbd> |
+
+### STD430
+| Name             | Bytes                                                                                                                      |
+|------------------|----------------------------------------------------------------------------------------------------------------------------|
+| mat2x2\<fp16>[]  | <kbd>[bbbb][bbbb]                                                                                                   </kbd> |
+| mat2x2\<float>[] | <kbd>[bbbb][bbbb][bbbb][bbbb]                                                                                       </kbd> |
+| mat2x3\<float>[] | <kbd>[bbbb][bbbb][bbbb][----]<br/>[bbbb][bbbb][bbbb][----]                                                          </kbd> |
+| mat2x4\<float>[] | <kbd>[bbbb][bbbb][bbbb][bbbb]<br/>[bbbb][bbbb][bbbb][bbbb]                                                          </kbd> |
+| mat2x2\<int8>[]  | <kbd>[bbbb]                                                                                                         </kbd> |
+| mat3x2\<float>[] | <kbd>[bbbb][bbbb][bbbb][bbbb]<br/>[bbbb][bbbb]                                                                      </kbd> |
+| mat3x3\<float>[] | <kbd>[bbbb][bbbb][bbbb][----]<br/>[bbbb][bbbb][bbbb][----]<br/>[bbbb][bbbb][bbbb][----]                             </kbd> |
+| mat4x2\<float>[] | <kbd>[bbbb][bbbb][bbbb][bbbb]<br/>[bbbb][bbbb][bbbb][bbbb]                                                          </kbd> |
+| mat4x3\<float>[] | <kbd>[bbbb][bbbb][bbbb][----]<br/>[bbbb][bbbb][bbbb][----]<br/>[bbbb][bbbb][bbbb][----]<br/>[bbbb][bbbb][bbbb][----]</kbd> |
+
+## Structures
+```
+struct {
+  int32 w;
+  float x;
+}
+```
+
+The STD140 pads 8 bytes at the end to become a multiple of 16 bytes.
```

Comment:
But that extra padding only is used if this struct is the element of an array, and that multiple-of-16bytes is the array's stride.

For example, if this struct is nested as the first member of an enclosing struct, and that enclosing struct had a second member which was a float y, then even in std140 that y would be placed at overall offset 8, i.e. within what would normally be considered padding of the inner struct (if you get what I mean).

Reply:
I added this example as another entry, it doesn't seem to pad the way you expect, heh.

Reply:
> The Offset decoration of a member must not place it between the end of a structure or an array and the next multiple of the alignment of that structure or array.

That rule is why it gets padded.